scenes/performance.gd

```diff
@@ -41,68 +41,113 @@ func print_song_loading_debug(to_print):
-#			for note in level.notes:
-#				print_song_loading_debug("---note -" + str(note)  + "- time-" + str(note.time)) #TODO expand
-#
-#			print_song_loading_debug("--- end notes data ----")
-#
-#		print_song_loading_debug("--- end level data ----")
+		audio_stream.stream = current_stream
+		audio_stream.play()
+	else:
+		print("No song selected")
+
+func sync_audiostream_remote(p_playing : bool, p_seek):
+	rpc("_sync_audiostream_remote", p_playing, p_seek, -1)
+
+@rpc("any_peer", "call_local", "reliable") func _sync_audiostream_remote(p_playing : bool, p_seek : float, relay_source):
+	if(multiplayer.get_unique_id() == relay_source):
+		return
+	if(audio_stream.playing == p_playing):
+		# do nothing.
+		return
+	else:
+		if(p_playing):
+			audio_stream.play(p_seek)
+		else:
+			audio_stream.stream_paused = true
+			if multiplayer.get_unique_id() != multiplayer.get_remote_sender_id():
+				# relay pause command to sychronize state.
+				rpc("_sync_audiostream_remote", p_playing, p_seek, multiplayer.get_remote_sender_id())
+			
 		
-#		var tab_creator = preload("res://scenes/tab_creator/tab_creator.gd").new()
-#		tab_creator.create_tabs(current_song)
+func _ready():
+	SessionVariables.song_changed.connect(_on_song_changed)
+	
+	current_song = SessionVariables.current_song
+	print_song_loading_debug(current_song)
 		
-		match SessionVariables.instrument:
-			SessionVariables.Instrument.GUITAR:
-				performance_instrument = preload("res://scenes/performance/instruments/performance_guitar.tscn").instantiate()
-			SessionVariables.Instrument.PIANO:
-				performance_instrument = preload("res://scenes/performance/instruments/performance_piano.tscn").instantiate()
-			SessionVariables.Instrument.PHIN:
-				pass
+	if current_song == null && SessionVariables.load_test_song:
+		var current_stream = load_mp3_from_path("res://Arlow - How Do You Know [NCS Release].mp3")
+		audio_stream.stream = current_stream
+		audio_stream.play()
 		
-		add_child(performance_instrument)
-		performance_instrument.start_game(current_song)
+	match SessionVariables.instrument:
+		SessionVariables.Instrument.GUITAR:
+			performance_instrument = preload("res://scenes/performance/instruments/performance_guitar.tscn").instantiate()
+		SessionVariables.Instrument.PIANO:
+			performance_instrument = preload("res://scenes/performance/instruments/performance_piano.tscn").instantiate()
+		SessionVariables.Instrument.PHIN:
+			pass
 	
-	audio_stream.stream = current_stream
-	audio_stream.play()
+	if (!is_instance_valid(performance_instrument)):
+		push_error("Invalid performance instrument")
+		# TODO : add error handling.
+	add_child(performance_instrument)
+	
+	_on_song_changed()
 #	audio_stream.volume_db = -16
 
+func is_playing() -> bool:
+	return audio_stream.playing
 
 func _process(delta: float) -> void:
 	if Input.is_action_pressed("ui_cancel"):
 		get_tree().change_scene_to_file("res://scenes/instrument_selection/instrument_selection.tscn")
 		return
-
-
+	else:
```

Comment:
No need for this else branch; return on the line before takes care of that.

Reply:
It's more like coding preference. You could as well keep the else and remove the return.
Although in here they do the same thing they have different meanings.
The else in here is the common method of flow control. The return means termination of execution of any more code.
I don't approve of the practice of using return statements for flow control. It is a bit confusing.
The return in that portion of the code shouldn't be treated as flow control but rather as a precaution. When you change the scene you shouldn't attempt to use the existing nodes. Because they might not be in a accessible state right after you change. And the return statement is supposed to prevent any funny business if the code below changes.
I tried to explain the idea to the best of my ability. Not sure if you got it's meaning.

Reply:
Yeah I get what you're saying, think it's really dependant on what you're used to. I prefer return statements over else statements in most cases because you prevent nested branches which can get really hard to read after a while ("wait, am I still in that one if-statement?"). I agree that there are cases where readability would be aided by an else statement, but in this case I wouldn't say that's the case. You've got a unique if-case with just a simple, single command right after. Should I still execute? -> If no, return. The rest of the code is in no way related to that if-statement, so I wouldn't use else here personally.

I _would_ use else with return in cases where the if has a clear true and false case that are very related to each other. I wouldn't use it when you're doing checks just to see if the rest of the function should still be executed.